File: history/valid_parantheses.py
class Solution:
    def isValid(self, s):
        n = len(s)
        stack = ['_']
        match = {')' : '(', '}' : '{', ']' : '['}

        for c in s:
            if c in match.keys():
                if stack.pop() != match[c]:
                    return False
            else:
                stack.append(c)

        return len(stack) == 1
        

File: history/valid-parantheses.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
public:
  bool match(char left, char right){
    if (left == '(' && right == ')') return true;
    else if (left == '[' && right == ']') return true;
    else if (left == '{' && right == '}') return true;
    else return false;
  }

  bool isValid(string s){
    stack<char> parantheses;
    if (s.length() % 2 != 0)
      return false;
    for (auto c : s){
      if (parantheses.empty())
        parantheses.push(c);
      else if (match(parantheses.top(),c))
        parantheses.pop();
      else
        parantheses.push(c);
    }
    return parantheses.empty();
  }
};
